### PR TITLE
oc tag should retry on conflict errors

### LINKED ISSUE #5526: Test flake: oc tag can fail with imageStream resource conflict
Looks like another test flake resulting from concurrent imageStream updates which is slightly different than the ISM fix in #5455.

```
Waiting for the import to complete, CTRL+C to stop waiting.
The import completed successfully.

Error from server: imageStream "newrepo" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
!!! Error in test/cmd/images.sh:110
	'oc tag mysql:5.5 newrepo:latest' exited with status 1
Call stack:
	1: test/cmd/images.sh:110 main(...)
Exiting with status 1
!!! Error in hack/test-cmd.sh:276
	'${test}' exited with status 1
Call stack:
	1: hack/test-cmd.sh:276 main(...)
Exiting with status 1
make: *** [test-int-plus] Error 1
```

@smarterclayton suggested we might need the same sort of idempotence fix in the imageStream REST handler.

### PR DESCRIPTION
**Pull request by smarterclayton**
Tagging is an overwrite.  Also update output to be cleaner.

Fixes #5526

### CODE REVIEW

**Comment by smarterclayton** [comment]
[test]

File: pkg/cmd/cli/cmd/tag.go [hunk 1]
```diff
@@ -289,74 +293,93 @@ func (o TagOptions) RunTag() error {
...
-			target.Spec.Tags[destTag] = targetRef
-			msg = fmt.Sprintf("Tag %s set up to track tag %s/%s.", o.ref.Exact(), o.destNamespace[i], destNameAndTag)
-		} else {
-			// The user wants to delete a spec tag.
-			if _, ok := target.Spec.Tags[destTag]; !ok {
-				return fmt.Errorf("destination tag %s/%s does not exist.\n", o.destNamespace[i], destNameAndTag)
+			// Check the stream creation timestamp and make sure we will not
+			// create a new image stream while deleting.
+			if target.CreationTimestamp.IsZero() && !o.deleteTag {
+				_, err = isc.Create(target)
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/tag.go]: not uid or resourceVersion?

Comment (liggitt) [inline pkg/cmd/cli/cmd/tag.go]: nm, I see it's a move of existing code... still weird

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/tag.go]: Eh.

On Nov 3, 2015, at 1:09 AM, Jordan Liggitt <notifications@github.com> wrote:

In pkg/cmd/cli/cmd/tag.go
<https://github.com/openshift/origin/pull/5626#discussion_r43718779>:

>  			}
>
> -			target.Spec.Tags[destTag] = targetRef
> -			msg = fmt.Sprintf("Tag %s set up to track tag %s/%s.", o.ref.Exact(), o.destNamespace[i], destNameAndTag)
> -		} else {
> -			// The user wants to delete a spec tag.
> -			if _, ok := target.Spec.Tags[destTag]; !ok {
> -				return fmt.Errorf("destination tag %s/%s does not exist.\n", o.destNamespace[i], destNameAndTag)
> +			// Check the stream creation timestamp and make sure we will not
> +			// create a new image stream while deleting.
> +			if target.CreationTimestamp.IsZero() && !o.deleteTag {
> +				_, err = isc.Create(target)

not uid or resourceVersion?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5626/files#r43718779>.

**Comment by kargakis** [comment]
How about using PATCH?

**Comment by soltysh** [comment]
@smarterclayton I don't like that big chunk of code inside the `kclient.RetryOnConflict` I'd prefer to move that anonymous internal function into a separate one. This will greatly improve readability.

File: pkg/cmd/cli/cmd/tag.go [hunk 2]
```diff
@@ -289,74 +293,93 @@ func (o TagOptions) RunTag() error {
...
-			localRef := o.ref
-			switch o.sourceKind {
-			case "DockerImage":
-				targetRef.From.Name = localRef.String()
-			default:
-				targetRef.From.Name = localRef.NameString()
-				targetRef.From.Namespace = o.ref.Namespace
+				sameNamespace := o.namespace == o.destNamespace[i]
+				target.Spec.Tags[destTag] = targetRef
+				if o.aliasTag {
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/tag.go]: exercise the messages in test-cmd to make sure we're seeing what we expect

**Comment by deads2k** [comment]
Why aren't we resolving this one server-side like we did in ISM?

**Comment by smarterclayton** [comment]
Because this is a client side operation and we have no server side endpoint.

I don't want to make a larger refactor at this point - I want the bare
minimum to get this working.

**Comment by liggitt** [comment]
+1 to method, post-3.1

**Comment by deads2k** [comment]
> this is a client side operation and we have no server side endpoint.

ick.  https://github.com/openshift/origin/issues/5637

**Comment by smarterclayton** [comment]
Test case added, please rereview.  Still on the bubble.

**Comment by liggitt** [comment]
code LGTM

**Comment by smarterclayton** [comment]
[merge]

**Pull request merged**